### PR TITLE
Fix runner deadlock

### PR DESCRIPTION
**Pull request by FinchPowers**

### CODE REVIEW

File: service/runner.cc [hunk 1]
```diff
@@ -388,6 +388,15 @@ run(const vector<string> & command,
...
     runImpl(command, onTerminate, stdOutSink, stdErrSink);
+    int counter = 1;
+    while (childPid_ == -1) {
+        ML::sleep(0.1);
+        ++ counter;
+        if (counter % 50 == 0) {
+            cerr << "Runner waiting after child...\n";
+            counter = 1;
+        }
+    }
```
Comment (FinchPowers) [inline service/runner.cc]: May not be needed. Let's discuss it.

Comment (wsourdeau) [inline service/runner.cc]: What is the bug that this patch is trying to fix?

Comment (FinchPowers) [inline service/runner.cc]: A deadlock in the runner.

File: service/runner.cc [hunk 2]
```diff
@@ -90,9 +90,14 @@ CreateStdPipe(bool forWriting)
...
 
+
 namespace Datacratic {
 
-/* ASYNCRUNNER */
+/****************************************************************************/
+/* RUNNER                                                                   */
+/****************************************************************************/
+
+std::string Runner::runnerHelper;
```
Comment (FinchPowers) [inline service/runner.cc]: If I understand correctly, this var is only there for testing? It should be indicated.

File: service/runner.cc [hunk 3]
```diff
@@ -346,7 +348,7 @@ attemptTaskTermination()
             cerr << "childPid_ >= 0\n";
         }
         if (!(task_.statusState == ProcessState::STOPPED
-              || task_.statusState == DONE)) {
+              || task_.statusState == ProcessState::DONE)) {
```
Comment (FinchPowers) [inline service/runner.cc]: Nice, we maintain comments. Put them within an if (0) to make them compile if they need to be maintained.

File: service/runner.cc (hunk 2 above)
Comment (FinchPowers) [inline service/runner.cc]: I see it was done in the .h file. Nevermind then.

File: service/testing/runner_test.cc [hunk 4]
```diff
@@ -616,3 +619,22 @@ BOOST_AUTO_TEST_CASE( test_set_prctl_from_thread )
     BOOST_CHECK_EQUAL(runResult.returnCode, 0);
     BOOST_CHECK_EQUAL(runResult.signum, -1);
 }
+#endif
+
+#if 1
+/* This test ensures that onTerminate is called with the appropriate RunResult
+ * when the runWrapper process fails. */
+BOOST_AUTO_TEST_CASE( test_unexisting_runner_helper )
```
Comment (FinchPowers) [inline service/testing/runner_test.cc]: Can you confirm that prior to the fix the deadlock was reproduced with that test?

**Comment by FinchPowers** [comment]
+1 if the test reproduced the deadlock.

File: service/testing/runner_test.cc (hunk 4 above)
Comment (wsourdeau) [inline service/testing/runner_test.cc]: I can.

File: service/runner.cc (hunk 3 above)
Comment (wsourdeau) [inline service/runner.cc]: I'll comment the block, explaining why it has to stay there.

File: service/runner.cc (hunk 2 above)
Comment (wsourdeau) [inline service/runner.cc]: Yes, it was.

File: service/testing/runner_test.cc (hunk 4 above)
Comment (wsourdeau) [inline service/testing/runner_test.cc]: And btw, of course it does. Otherwise the test would be useless.

**Comment by wsourdeau** [comment]
Squash merged

**Pull request closed without merging**